Library/Homebrew/cmd/upgrade.rb

```diff
@@ -104,6 +105,35 @@ def upgrade
         onoe "#{f}: #{e}"
       end
     end
+
+    unless Homebrew.failed?
+      #  Assess existing kegs and reinstall any broken.
+      broken_formulae = []
+
+      CacheStoreDatabase.use(:linkage) do |db|
+        kegs = Formula.installed.collect(&:opt_or_installed_prefix_keg).reject(&:nil?)
+
+        odebug "Assessing keg linkage status" if kegs.size > 1
+
+        kegs.each do |keg|
+          checker = LinkageChecker.new(keg, cache_db: db)
+          next unless checker.broken_library_linkage?
+
+          odebug "#{formula} has broken dynamic library links"
+
+          broken_formulae << resolve_formula(keg)
+        end
+      end
+
+      return if broken_formulae.empty?
+
+      opoo "Reinstalling #{Formatter.pluralize(broken_formulae.length, "keg")} with broken dynamic library links"
+
+      broken_formulae.each do |f|
+        reinstall_formula(f)
```

Comment:
do we want to consider a reinstall with `--build-from-source`? if it is just going to pour a broken bottle again, then it won't fix anything

Reply:
Agreed to building from source here 👍

Reply:
This should use `fi = FormulaInstaller.new(f)` rather than including `cmd/reinstall` here.